### PR TITLE
refactor(compute-oslogin): Rewriting the OSLogin SSH sample and it's test

### PR DESCRIPTION
**Pull request by m-strzelczyk**
## Description

Issue #8429 

Rewriting the SSH sample and its test using new libraries and hopefully avoiding the nasty test flakiness.

This commit contains the new sample and test for it. It only disables (by skipping) the test for old sample. The old sample is for now left as it is, because we first need to change the sample used on [this documentation page](https://cloud.google.com/compute/docs/tutorials/service-account-ssh).

### CODE REVIEW

File: compute/oslogin/new_service_account_ssh.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    if returncode:
+        print(f'Command returned error status {returncode}')
+        if capture_output:
+            print(f"With output: {output}")
+
+    return returncode, output
+
+
+def create_ssh_key(oslogin_client: oslogin_v1.OsLoginServiceClient,
+                   account: str, expire_time: int = 300):
```
Comment (dandhlee) [inline compute/oslogin/new_service_account_ssh.py:71]: missing return type hint

File: compute/oslogin/new_service_account_ssh.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        account: Name of the service account this key will be assigned to.
+            This should be in form of `user/<service_account_username>`.
+        expire_time: How many seconds from now should this key be valid.
+
+    Returns:
+        The path to private SSH key. Public key can be found by appending `.pub`
+        to the file name.
+
+    """
+    private_key_file = '/tmp/key-' + str(uuid.uuid4())
```
Comment (dandhlee) [inline compute/oslogin/new_service_account_ssh.py:86]:
```suggestion
    private_key_file = f'/tmp/key-{str(uuid.uuid4()}'
```

File: compute/oslogin/new_service_account_ssh.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    execute(['ssh-keygen', '-t', 'rsa', '-N', '', '-f', private_key_file])
+
+    with open(private_key_file + '.pub', 'r') as original:
```
Comment (dandhlee) [inline compute/oslogin/new_service_account_ssh.py:89]:
```suggestion
    with open(f'{private_key_file}.pub', 'r') as original:
```

File: compute/oslogin/new_service_account_ssh.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    print(f'Setting key for {account}...')
+    oslogin_client.import_ssh_public_key(request)
+
+    # Let the key properly propagate
+    time.sleep(5)
+
+    return private_key_file
+
+
+def run_ssh(cmd: str, private_key_file: str, username: str, hostname: str):
```
Comment (dandhlee) [inline compute/oslogin/new_service_account_ssh.py:109]: missing return type hint

File: compute/oslogin/new_service_account_ssh.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    ssh = subprocess.run(
+        ssh_command, shell=False, stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT, text=True, check=True,
+        env={'SSH_AUTH_SOCK': ''})
+
+    return ssh.stdout
+
+
+def main(cmd, project: str, instance: Optional[str] = None, zone: Optional[str] = None,
+         account: Optional[str] = None, hostname: Optional[str] = None, oslogin: Optional[oslogin_v1.OsLoginServiceClient] = None):
```
Comment (dandhlee) [inline compute/oslogin/new_service_account_ssh.py:137]:
```suggestion
def main(
    cmd,
    project: str,
    instance: Optional[str] = None,
    zone: Optional[str] = None,
    account: Optional[str] = None,
    hostname: Optional[str] = None,
    oslogin: Optional[oslogin_v1.OsLoginServiceClient] = None
) -> None:
```

File: compute/oslogin/new_service_account_ssh.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    # Create the OS Login API object.
+    if oslogin is None:
+        oslogin = oslogin_v1.OsLoginServiceClient()
+
+    # Identify the service account ID if it is not already provided.
+    account = account or requests.get(
+        SERVICE_ACCOUNT_METADATA_URL, headers=HEADERS).text
+
+    if not account.startswith('users/'):
+        account = 'users/' + account
```
Comment (dandhlee) [inline compute/oslogin/new_service_account_ssh.py:149]:
```suggestion
        account = f'users/{account}'
```

File: compute/oslogin/new_service_account_ssh.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+        # Run a command on the remote instance over SSH.
+        result = run_ssh(cmd, private_key_file, username, hostname)
+
+        # Print the command line output from the remote instance.
+        print(result)
+    finally:
+        # Shred the private key and delete the pair.
+        execute(['shred', private_key_file])
+        execute(['rm', private_key_file])
+        execute(['rm', private_key_file + '.pub'])
```
Comment (dandhlee) [inline compute/oslogin/new_service_account_ssh.py:173]:
```suggestion
        execute(['rm', f'{private_key_file}.pub'])
```

**Review by dandhlee — CHANGES_REQUESTED** [review]

File: compute/oslogin/new_service_account_ssh_test.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+                        'role': 'roles/iam.serviceAccountUser'
+                    }
+                ]
+            }
+        }).execute()
+
+    # Create a service account key.
+    service_account_key = iam.projects().serviceAccounts().keys().create(
+        name=f'projects/{PROJECT}/serviceAccounts/{account_email}',
+        body={}
```
Comment (dandhlee) [inline compute/oslogin/new_service_account_ssh_test.py:82]:
```suggestion
        body={},
```

File: compute/oslogin/new_service_account_ssh_test.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+    yield {'name': account_email, 'credentials': credentials}
+
+    # Cleanup
+    # Delete the temporary service account and its associated keys.
+    try:
+        iam.projects().serviceAccounts().delete(
+            name=f'projects/{PROJECT}/serviceAccounts/{account_email}'
+        ).execute()
+    except Exception:
```
Comment (dandhlee) [inline compute/oslogin/new_service_account_ssh_test.py:98]: Are there specific exceptions you can use?

File: compute/oslogin/new_service_account_ssh_test.py [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+    # Wait for everything to propagate
+    time.sleep(5)
+
+    yield client.get(project=PROJECT, zone=ZONE, instance=instance.name)
+
+    client.delete(project=PROJECT, zone=ZONE, instance=instance.name).result()
+
+
+def test_oslogin_ssh(oslogin_instance, oslogin_service_account, capsys):
+    account = 'users/' + oslogin_service_account['name']
```
Comment (dandhlee) [inline compute/oslogin/new_service_account_ssh_test.py:193]:
```suggestion
    account = f'users/{oslogin_service_account['name']}'
```

File: compute/oslogin/new_service_account_ssh_test.py [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+    oslogin_client = oslogin_v1.OsLoginServiceClient(credentials=oslogin_service_account['credentials'])
+    # Letting everything settle down...
+    time.sleep(60)
```
Comment (dandhlee) [inline compute/oslogin/new_service_account_ssh_test.py:196]: Should we use a try-except afterwards in case this time wasn't long enough?

Comment (m-strzelczyk) [reply compute/oslogin/new_service_account_ssh_test.py:197]: If this time was not long enough, then I think we have a problem with the resource setup. The code already waits for completion of operations like instance creation. IMO, 60 seconds is quite a safe margin.

File: compute/oslogin/new_service_account_ssh.py [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+
+SERVICE_ACCOUNT_METADATA_URL = (
+    'http://metadata.google.internal/computeMetadata/v1/instance/'
+    'service-accounts/default/email')
+HEADERS = {'Metadata-Flavor': 'Google'}
+
+
+def execute(cmd: List[str], cwd: Optional[str] = None,
+            capture_output: Optional[bool] = False, env: Optional[dict] = None,
+            raise_errors: Optional[bool] = True) -> Tuple[int, str]:
```
[the remark this answers is not in the document]
Comment (m-strzelczyk) [reply compute/oslogin/new_service_account_ssh.py]: Sure, I get that, but `list[str]` is not legal, while `List[str]` is :)

[the remark this answers is not in the document]
Comment (dandhlee) [reply compute/oslogin/new_service_account_ssh.py]: Ah you're right. Yes, please keep them as `List`. Sorry about that!

File: compute/oslogin/new_service_account_ssh_test.py (hunk 11 above)
Comment (dandhlee) [reply compute/oslogin/new_service_account_ssh_test.py:197]: SGTM.

**Review by dandhlee — APPROVED** [review]
Thank you!

**Review by engelke — APPROVED** [review]

**Pull request merged**